crates/workers/src/workers.rs

```diff
@@ -0,0 +1,571 @@
+use parking_lot::RwLock;
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+pub struct WorkerInfo {
+    pub deal_id: String,
+    pub creator: PeerId,
+    pub active: RwLock<bool>,
+}
+
+pub struct Workers {
+    /// deal_id -> worker_id
+    worker_ids: RwLock<HashMap<DealId, WorkerId>>,
+    /// worker_id -> worker_info
+    worker_infos: RwLock<HashMap<WorkerId, WorkerInfo>>,
+    workers_dir: PathBuf,
+
+    key_storage: Arc<KeyStorage>,
+    scopes: Scopes,
+}
+
+impl Workers {
+    pub fn new(key_storage: Arc<KeyStorage>, scopes: Scopes) -> Self {
+        Self {
+            worker_ids: Default::default(),
+            worker_infos: Default::default(),
+            workers_dir: Default::default(),
+            key_storage,
+            scopes,
+        }
+    }
+
+    pub async fn from_path(
+        workers_dir: &Path,
+        key_storage: Arc<KeyStorage>,
+        scopes: Scopes,
+    ) -> eyre::Result<Self> {
+        let workers = load_persisted_workers(workers_dir).await?;
+        let mut worker_ids = HashMap::with_capacity(workers.len());
+        let mut worker_infos = HashMap::with_capacity(workers.len());
+
+        for w in workers {
+            let worker_id = w.worker_id;
+            let deal_id = w.deal_id.clone();
+            worker_infos.insert(worker_id, w.into());
+            worker_ids.insert(deal_id, worker_id);
+        }
+        Ok(Self {
+            worker_ids: RwLock::new(worker_ids),
+            worker_infos: RwLock::new(worker_infos),
+            workers_dir: workers_dir.to_path_buf(),
+            key_storage,
+            scopes,
+        })
+    }
+
+    pub async fn create_worker(
+        &self,
+        deal_id: String,
+        init_peer_id: PeerId,
+    ) -> Result<PeerId, WorkersError> {
+        let worker_id = {
+            let guard = self.worker_ids.read();
+            guard.get(&deal_id).cloned()
+        };
+        match worker_id {
+            Some(_) => Err(WorkersError::WorkerAlreadyExists { deal_id }),
+            _ => {
+                let key_pair = self
+                    .key_storage
+                    .create_key_pair()
+                    .await
+                    .map_err(|err| WorkersError::CreateWorkerKeyPair { err })?;
+
+                let worker_id = key_pair.get_peer_id();
+
+                match self
+                    .store_worker(worker_id, deal_id.clone(), init_peer_id)
+                    .await
+                {
+                    Ok(worker_info) => {
+                        let mut worker_ids = self.worker_ids.write();
+                        let mut worker_infos = self.worker_infos.write();
+
+                        if worker_ids.contains_key(&deal_id) {
+                            return Err(WorkersError::WorkerAlreadyExists { deal_id });
+                        }
+
+                        worker_ids.insert(deal_id, worker_id);
+                        worker_infos.insert(worker_id, worker_info);
```

Comment:
maybe make locks as short as possible?

```suggestion
                        let mut worker_ids = self.worker_ids.write();

                        if worker_ids.contains_key(&deal_id) {
                            return Err(WorkersError::WorkerAlreadyExists { deal_id });
                        }

                        worker_ids.insert(deal_id, worker_id);
                        drop(worker_ids);
                        
                        let mut worker_infos = self.worker_infos.write();
                        worker_infos.insert(worker_id, worker_info);
```

Reply:
It is a potential race, so lets keep it